FHEM/14_SD_UT.pm

```diff
@@ -1121,13 +1121,12 @@ sub SD_UT_Parse($$) {
 			Log3 $iohash, 4, "$ioname: SD_UT device - RC_10 button: $button | state: $state";
 			
 			### if receive device _all, set A | B | C | D ###
-			my $device_all = "RC_10 ".$deviceCode."_all";
-			if ($devicedef =~ /^$device_all$/) {
-				foreach my $d (sort keys %defs) {
-					if (defined($defs{$d}) && defined($defs{$d}{NAME}) && $defs{$d}{NAME} =~ /^RC_10_$deviceCode.[ABCD]$/) {
-						readingsSingleUpdate($defs{$d}, "state" , $state , 1);
-						Log3 $iohash, 4, "$ioname: SD_UT device - ".$defs{$d}{NAME}." | state: $state";
-					}
+			my @list = devspec2array("DEF=RC_10 $deviceCode"."_.*");
```

Comment:
Ich hab mir das noch ein bisschen durch den Kopf gehen lassen.
Ursprünglich hatte ich angenommen, dass eine Suche nach TYPE die beste Wahl sein dürfte.

Das scheint aber nicht unbedingt der Fall zu sein, da devspec2array alle Devices durchsucht. Allerdings können wir das _all Device direkt durch einen Filter ausschließen.

`my @list = devspec2array("DEF=RC_10 $deviceCode"."_.*,:FILTER=DEF!=RC_10 $deviceCode"."_all");`

Reply:
Nur so eine Idee:
Warum überhaupt erst danach suchen und in ein Array packen?
Die 4 möglichen Definitionen der Geräte sind ja bekannt: RC_10 $deviceCode"."_. A|B|C|D
Da kann man doch gleich ein
readingsSingleUpdate($defs{$_}, "state" , $state , 1); if (defined RC_10 $deviceCode"."_. A|B|C|D)
in einer Schleife laufen lassen.

Reply:
Wenn die Bezeichnung der Definition exakt bekannt ist, dann geht das natürlich auch so.

Reply:
Was haltet Ihr von der Variante. Es kommt auf das gleiche wie mit einer Schleife heraus und umfasst nur die nötigsten Codezeilen. (nur die 4 letzten Zeilen)

```
if ($button eq "all") {
	$state = substr($bitData,20,4);
	if ($state eq "0100") {
		$state = "on";
	} elsif ($state eq "1000") {
		$state = "off";
	} else {
		$state = "unknown";
		$def = undef;
	}

	### if receive device _all, set A | B | C | D ###
	readingsSingleUpdate("RC_10 " . $deviceCode ."_A", "state" , $state , 1) if (defined $defs{"RC_10 " . $deviceCode ."_A"});
	readingsSingleUpdate("RC_10 " . $deviceCode ."_B", "state" , $state , 1) if (defined $defs{"RC_10 " . $deviceCode ."_B"});
	readingsSingleUpdate("RC_10 " . $deviceCode ."_C", "state" , $state , 1) if (defined $defs{"RC_10 " . $deviceCode ."_C"});
	readingsSingleUpdate("RC_10 " . $deviceCode ."_D", "state" , $state , 1) if (defined $defs{"RC_10 " . $deviceCode ."_D"});
}
```

Reply:
Ja, so meinte ich das. Ob es noch Sinn macht, das in eine "for" zu packen, kann ich nicht einschätzen. Ich schätze, bei 4 Zeilen ist das noch vertretbar. Viel Code ist schneller :-)

Reply:
Ja, kann man so machen.
Über eine Schleife wären es drei Zeilen glaube ich.
Wenn tatsächlich nur das Reading starte aktualisiert werden muss, kann man es so lassen.